Cant restart the game

```
[serve-watch] Error: ID 1 is already taken (by a player).
[serve-watch]     at ServerSimulation.Simulation.addEnemy (/home/erik/Documents/vov/src/common/simulation.ts:83:13)
[serve-watch]     at ServerSimulation.Simulation.commonUpdate (/home/erik/Documents/vov/src/common/simulation.ts:54:12)
[serve-watch]     at ServerSimulation.update (/home/erik/Documents/vov/src/server/serverSimulation.ts:14:10)
[serve-watch]     at ServerGame.doUpdate (/home/erik/Documents/vov/src/server/game.ts:75:21)
[serve-watch]     at ServerGame.GameLoop.update (/home/erik/Documents/vov/src/common/game-loop.ts:65:12)
[serve-watch]     at Timeout._onTimeout (/home/erik/Documents/vov/src/common/game-loop.ts:110:25)
[serve-watch]     at listOnTimeout (node:internal/timers:556:17)
[serve-watch]     at processTimers (node:internal/timers:499:7)
[serve-watch] [ERROR] 14:01:48 Error: ID 1 is already taken (by a player).
```

when restarting the game second time

--- a/public/index.html
+++ b/public/index.html
@@ -7,6 +7,8 @@
 </head>
 <body>
   <script type="text/javascript" src="bundle.js"></script>
-  <div id="game-container"></div>
+  <div id="game-container">
+    <button id="btn-ready">:-(</button>
+  </div>
 </body>
 </html>

--- a/src/client/clientSimulation.ts
+++ b/src/client/clientSimulation.ts
@@ -6,8 +6,8 @@ import State from '../common/state';
 import { Body } from 'planck-js';
 
 export default class ClientSimulation extends Simulation {
-  constructor(map: GameMap, updateStep: number, enemyIdCounter: number) {
-    super(map, updateStep, enemyIdCounter);
+  constructor(map: GameMap, updateStep: number, numPlayers: number) {
+    super(map, updateStep, numPlayers);
   }
 
   public update(body: Body, input: Input): void {

--- a/src/client/main.ts
+++ b/src/client/main.ts
@@ -65,6 +65,11 @@ function setup(): void {
     channel.onDisconnect(_reason => {
       game?.stop();
     });
+
+    // TODO: accept user input
+    setTimeout(() => {
+      channel.emit('ready', { status: true }, { reliable: true });
+    }, 1000);
   });
 }
 

--- a/src/client/main.ts
+++ b/src/client/main.ts
@@ -10,9 +10,18 @@ import {
 } from '../common/constants';
 import GameMap from '../common/gameMap';
 
+function onDocumentReady(callback: () => void): void {
+  // https://codetonics.com/javascript/detect-document-ready/
+  if (document.readyState !== 'loading') {
+    callback();
+  } else {
+    document.addEventListener('DOMContentLoaded', callback);
+  }
+}
+
 PIXI.settings.SCALE_MODE = PIXI.SCALE_MODES.NEAREST;
 
-function setup(): void {
+function finishedResources(): void {
   const channel = geckos({ port: PORT });
   const renderer = PIXI.autoDetectRenderer();
   const stage = new PIXI.Container();
@@ -23,6 +32,7 @@ function setup(): void {
 
   const { maxMessageSize } = channel;
   let game;
+  let readyStatus = false;
 
   channel.onConnect(error => {
     if (error) {
@@ -33,6 +43,11 @@ function setup(): void {
     channel.onRaw(data => game?.serverMsg(data));
 
     channel.on('start', data => {
+      const button = document.getElementById('btn-ready');
+      if (button !== null) {
+        button.remove();
+      }
+
       const map = new GameMap(data['map'], data['tileset']);
       const [
         maps_total_pixel_width,
@@ -66,10 +81,18 @@ function setup(): void {
       game?.stop();
     });
 
-    // TODO: accept user input
-    setTimeout(() => {
-      channel.emit('ready', { status: true }, { reliable: true });
-    }, 1000);
+    onDocumentReady(() => {
+      const button = document.getElementById('btn-ready');
+
+      if (button !== null) {
+        button.addEventListener('click', () => {
+          if (game !== undefined) return;
+          readyStatus = !readyStatus;
+          button.innerText = ':-' + (readyStatus ? ')' : '(');
+          channel.emit('ready', { status: readyStatus }, { reliable: true });
+        });
+      }
+    });
   });
 }
 
@@ -77,4 +100,4 @@ PIXI.loader
   .add(PLAYER_SPRITE)
   .add(ENEMY_SPRITE)
   .add('imgs/tilesheets/scifitiles-sheet.png') // TODO: load from map somehow
-  .load(setup);
+  .load(finishedResources);

--- a/src/common/constants.ts
+++ b/src/common/constants.ts
@@ -92,9 +92,6 @@ export const SERVER_BROADCAST_RATE = 2;
  */
 export const INPUT_QUEUE_MAX_AGE = 3;
 
-/** The multiplier for the number of enemies to spawn in each wave. */
-export const WAVE_ENEMY_COUNT_INCREMENT = PLAYER_LIMIT;
-
 /** The increment in enemy health for each wave. */
 export const WAVE_ENEMY_HEALTH_INCREMENT = 2;
 

--- a/src/common/simulation.ts
+++ b/src/common/simulation.ts
@@ -16,6 +16,7 @@ export default abstract class Simulation {
   protected _stepCounter: number;
   protected _enemyIdCounter: number;
   protected _wave: Wave;
+  protected _numPlayers: number;
 
   public get map(): GameMap {
     return this._gameMap;
@@ -39,21 +40,18 @@ export default abstract class Simulation {
   /**
    * @param map The map to use
    * @param updateStep How big an update step is in seconds.
-   * @param enemyIdCounter The starting value for the enemy ids.
+   * @param numPlayers The number of players.
    */
-  constructor(map: GameMap, updateStep: number, enemyIdCounter: number) {
+  constructor(map: GameMap, updateStep: number, numPlayers: number) {
     this.updateStep = updateStep;
     this._world = this.createWorld(map);
     this._bodies = new Map<Id, Body>();
     this._state = new State();
     this._stepCounter = 0;
-    this._enemyIdCounter = enemyIdCounter;
+    this._enemyIdCounter = numPlayers;
     this._gameMap = map;
-    this._wave = new Wave(
-      1,
-      constants.WAVE_ENEMY_COUNT_INCREMENT,
-      constants.WAVE_ENEMY_HEALTH_INCREMENT,
-    );
+    this._wave = new Wave(1, numPlayers, constants.WAVE_ENEMY_HEALTH_INCREMENT);
+    this._numPlayers = numPlayers;
   }
 
   public commonUpdate(): void {
@@ -76,7 +74,7 @@ export default abstract class Simulation {
         constants.WAVE_COOLDOWN * constants.SERVER_UPS
       ) {
         const number = this._wave.waveNumber + 1;
-        const enemies = number * constants.WAVE_ENEMY_COUNT_INCREMENT;
+        const enemies = number * this._numPlayers;
         const health = number * constants.WAVE_ENEMY_HEALTH_INCREMENT;
         this._wave = new Wave(number, enemies, health);
         // TODO: remove this when we display the wave information on screen

--- a/src/server/index.ts
+++ b/src/server/index.ts
@@ -32,6 +32,11 @@ const player_list: Player[] = [];
 let game: ServerGame | undefined;
 
 function startGame(maxMessageSize?: number): void {
+  // normalize the players' IDs
+  for (let i = 0; i < player_list.length; i++) {
+    player_list[i].player_id = i;
+  }
+
   game = new ServerGame(
     new GameMap('scifi-1', 'scifi'),
     x => {
@@ -101,10 +106,7 @@ io.onConnection(channel => {
     if (index >= 0) {
       player_list[index].ready = status;
 
-      if (
-        player_list.length === PLAYER_LIMIT &&
-        player_list.every(p => p.ready)
-      ) {
+      if (player_list.every(p => p.ready)) {
         const { maxMessageSize } = channel;
         startGame(maxMessageSize);
       }

--- a/src/server/index.ts
+++ b/src/server/index.ts
@@ -24,6 +24,7 @@ type Player = {
   channel_id?: string;
   player_id: number;
   channel: ServerChannel;
+  ready: boolean;
 };
 
 let client_id = 0;
@@ -60,7 +61,11 @@ function startGame(maxMessageSize?: number): void {
 }
 
 io.onConnection(channel => {
+  if (game !== undefined) return;
+  if (player_list.length >= PLAYER_LIMIT) return;
+
   console.info(`${channel.id} connected`);
+
   channel.onDrop(drop => {
     console.warn('We are dropping packets: ', drop);
   });
@@ -71,6 +76,7 @@ io.onConnection(channel => {
     channel_id: channel.id,
     player_id: my_id,
     channel: channel,
+    ready: false,
   });
 
   channel.onDisconnect(() => {
@@ -80,15 +86,30 @@ io.onConnection(channel => {
       player_list.splice(i, 1);
       if (player_list.length === 0) {
         game?.stop();
+        client_id = 0;
+        game = undefined;
       }
     }
   });
 
-  // NOTE: temporary start condition
-  if (player_list.length === PLAYER_LIMIT) {
-    const { maxMessageSize } = channel;
-    startGame(maxMessageSize);
-  }
+  channel.on('ready', data => {
+    if (game !== undefined) return;
+
+    const status = data['status'];
+    const index = player_list.findIndex(p => p.player_id === my_id);
+
+    if (index >= 0) {
+      player_list[index].ready = status;
+
+      if (
+        player_list.length === PLAYER_LIMIT &&
+        player_list.every(p => p.ready)
+      ) {
+        const { maxMessageSize } = channel;
+        startGame(maxMessageSize);
+      }
+    }
+  });
 });
 
 if ((process.env.NODE_SERVER_TEST ?? '') === '') {

--- a/src/server/serverSimulation.ts
+++ b/src/server/serverSimulation.ts
@@ -5,8 +5,8 @@ import GameMap from '../common/gameMap';
 export default class ServerSimulation extends Simulation {
   public difficulty: number;
 
-  constructor(gameMap: GameMap, updateStep: number, enemyIdCounter: number) {
-    super(gameMap, updateStep, enemyIdCounter);
+  constructor(gameMap: GameMap, updateStep: number, numPlayers: number) {
+    super(gameMap, updateStep, numPlayers);
     this.difficulty = 0;
   }
 

